proposals/cache_v2.md

```diff
@@ -0,0 +1,177 @@
+
+#### Retrieve a layer digest by a selector
+
+The higher-level operation is finding a layer that was built by a list of layer entries (the files that built the layer). The operation generates a digest of the layer entries (the selector) and then the storage engine simply finds the file referenced at `selectors/<the selector digest>` (from which the layer digest is retrieved). This operation is *constant* in the size of the cache (and linear in the number of layer entries).
+
+### Directory sharding
+
+Directory traversal may be at worst a linear time operation (such as ext2 and FAT), and at best a logarithmic time operation (such as with the B-tree directory structure of ext3/4, HFS+, NTFS). The linear time directory traversal could potentially be an issue if the cache grows without bound. To address this issue, the contents of a directory can be sharded into a set number of bins to achieve logarithmic time traversal. For example, given a flat directory of files:
+
+```
+directory/
+  file1
+  file2
+  ...
+  file100000
+```
+
+The files can be placed in a select number of bins (say 4 in this case):
+
+```
+directory/
+  bin1/
+    file1
+    ...
+  bin2/
+    file25001
+    ...
+  bin3/
+    file50001
+    ...
+  bin4/
+    file75001
+    ...
+```
+
+This technique can be applied recursively within each bin as well (with a sentinel to denote recursive termination).
+
+By choosing a constant number of bins, the traversal at each directory is constant time. The total depth of the sharding is logarithmic to the number of files. *The overall traversal time is logarithmic.*
+
+In practice, an effective number of bins may be 256, named by the hex codes `00` through `ff`.
+
+### Locking
+
+All file writes will obtain an exclusive lock on the file and all file reads will obtain a shared lock on the file. However, obtaining locks on the same file during concurrent executions of Jib within the same JVM would result in an `OverlappingFileLockException`. Therefore, the cache cannot rely on `FileLock`s for concurrent I/O control on files.
+
+#### Save a cache entry
+
+Saving a cache entry requires saving the layer blob and metadata to a layer directory under `layers/` and the selector file to the `selectors/` directory. 
+
+The layer directory should be written atomically to avoid reads that may read an incomplete layer directory.
+
+The selector file can be written atomically anytime after the layer directory is fully written.
+
+All atomic writes should be done by first writing to a temporary location and then moving to the desired location.
+
+#### List the entries in the cache
+
+Listing the entries in the cache just requires obtaining a list of all the layer directories under `layers/`. Since the layer directory may write to a temporary directory within the `layers/` directory, the listing would just need to filter out those temporary directories.
```

Comment:
Perhaps we may avoid this filtering if we create a temp directory or file in other locations like `/tmp`?

Reply:
I believe on Windows moving a file from one directory to another is a costly operation (though I cannot find any sources for that currently). However, we could just go with creating the temp directory not in `layers/` and just absorb the cost on Windows for the cost we save during reads.